### ISSUE TITLE
No primary key constraint error when fetching PostgreSQL views

### ISSUE THREAD

**Issue by jgngo** [Budibase/budibase#6216]
**Hosting**
- Self
  - Method: docker compose
  - Budibase Version: 1.0.195

**Describe the bug**
When adding a PostgreSQL source and fetching the tables, an error occurs for the views.

Error fetching tables
No primary key constraint found for the following: <view_name>

PostgreSQL does not allow primary keys to be added to views.

This is the CREATE VIEW statement. The w."Id" is a primary key.

```
CREATE OR REPLACE VIEW public."ViewWarehouse"
AS SELECT w."Id",
    w."Name",
    w."ShortAddress",
    w."Street",
    w."Barangay",
    w."City",
    w."Town",
    w."FullAddress",
    w."Region",
    w."ZipCode",
    w."WarehouseTypeId",
    ( SELECT wt."Name" AS name
           FROM "WarehouseType" wt
          WHERE wt."Id" = w."WarehouseTypeId") AS "WarehouseType",
    w."Capacity",
    w."WarehouseUomId",
    ( SELECT w2."Alias" AS alias
           FROM "WarehouseUom" w2
          WHERE w2."Id" = w."WarehouseUomId") AS "UomAlias",
    ( SELECT w2."Name" AS name
           FROM "WarehouseUom" w2
          WHERE w2."Id" = w."WarehouseUomId") AS "UomName",
    w."StorageFee",
    w."HandingIn",
    w."HandingOut",
    w."Status",
    w."IsDelete",
    w."CreatedBy",
    w."CreatedDate",
    w."ModifiedBy",
    w."ModifiedDate"
   FROM "Warehouse" w;
```

**Comment by melohagan** [comment]
Hi @jgngo 

Currently primary keys are required for external relational datasources. 

So instead of using Postgres views, you can fetch the tables, and then create queries that return the same data that the view would.

**Comment by aviatrix** [comment]
OOOOOR... hear me out.. Add support for views. Views are in every RDBMS... it's silly not to support them

**Comment by melohagan** [comment]
@aviatrix Fair enough. I wasn't suggesting the idea was bad, I was just pointing out an alternative in the meantime.

**Comment by tvogt** [comment]
It is definitely needed not just for selection but also for updates. I want to build forms around views and can't. :-(

**Comment by garyboyes** [comment]
Would agree with aviatrix on this one, please add in support for views. The primary key constraint seems unnecessary given other similar tools (such as ReTool) do not have this constraint

**Comment by AlejandroFNadal** [comment]
I add my voice on this point, it is quite important to be able to read data from views. Perhaps this issue should be renamed to "Support Views - Currently getting 'No primary key' constraint error when fetching" since it does not work in MySQL as well as PostgreSQL. I was not able to find this issue and I created a duplicated one by mistake (sorry about that @melohagan )

**Comment by mintonan** [comment]
Just wanted to comment here as well that I have tables from older systems that do not have Primary keys, was there a work around someone found? I am running into an error with testing importing some tables where it will not let me add because there is no primary key

**Comment by tispratik** [comment]
We have UUIDs as primary keys, are those supported? We're getting the same error.

**Comment by daraul** [comment]
I suggest BB be updated to support a way of defining a "primary key" column for a table, or view.

**Comment by Marcello1099** [comment]
this is definitely relevant; specially with the filters constraints on queries.

**Comment by poirazis** [comment]
Would also like to raise my support for this feature !

**Comment by murphysprings** [comment]
This would be incredibly helpful to have this feature.

**Comment by mmorley0395** [comment]
This would be an incredibly useful feature!

**Comment by poirazis** [comment]
A rather easy workaround to this is for budibase to bring in views as queries ( that dont require a primary key )
if it detects a view , it creates a custom query select * from view and attaches is to the datasource 

so importing a view equals creating a query, even better with automatic offset / limit pagination

**Comment by mmorley0395** [comment]
@poirazis 
This is smart and a good suggestion for Budibase maintainers! And this line of thinking also gave me a temporary fix. I create all my views in Postgres, then, rather than creating duplicate queries (typing the whole query into budibase as a datasource), I just do 'select * from {view_name}' in the queries section of Budibase. This is helpful as I can manage all the views in Postgres and change them as needed, and as long as the names are the same, they can update in Budibase. Hopefully this helps someone else in the interim until this is addressed.

**Comment by melohagan** [comment]
Difficulty: Medium

We should let Budibase import DB views even when the JDBC metadata reports no primary key. A solution would be to let the user designate one or more columns as a logical key during the import flow (or infer from the VIEW definition), and skip the strict validation that currently rejects the view. That requires extending the fetcher to accept the override, storing the chosen key, and ensuring the runtime still treats the view as read-only so it won’t attempt inserts/updates without a real PK.

Although ideally you wouldn't need to select a pseudo-key
